packages/web-extension/src/walletManager/walletManagerWorker.ts

```diff
@@ -0,0 +1,103 @@
+ * Uses wallet and store factories to create wallets.
+ * Keeps track of created stores and reuses them when a wallet is reactivated.
+ */
+export class WalletManagerWorker implements WalletManagerApi {
+  #api$ = new BehaviorSubject<ObservableWallet | null>(null);
+  #hostSubscription: Shutdown;
+  #keyAgentSubscription: Shutdown;
+  #walletStores = new Map<string, storage.WalletStores>();
+
+  #walletFactory: WalletFactory;
+  #storesFactory: StoresFactory;
+  #logger: Logger;
+  #runtime: MinimalRuntime;
+
+  constructor({ walletFactory, storesFactory, logger, runtime }: MessengerDependencies & WalletManagerDependencies) {
+    this.#walletFactory = walletFactory;
+    this.#storesFactory = storesFactory;
+    this.#logger = logger;
+    this.#runtime;
+    this.#hostSubscription = exposeApi(
+      { api$: this.#api$.asObservable(), baseChannel: walletChannel, properties: observableWalletProperties },
+      { logger, runtime }
+    );
+  }
+
+  async activate({ walletName, cardanoServicesUrl }: WalletManagerActivateProps): Promise<void> {
+    if (await this.#isActive(walletName)) {
+      return; // already active
+    }
+    this.#deactivateWallet();
+
+    // Key agent is created by UI script on a per wallet unique channel
+    const keyAgent = consumeKeyAgent({ walletName }, { logger: this.#logger, runtime: this.#runtime });
+    this.#keyAgentSubscription = keyAgent;
+    const stores = this.#getStores(walletName);
+
+    // Wallet factory is responsible for creating the wallet and the providers based on cardanoServicesUrl
+    const wallet = await this.#walletFactory.create({ cardanoServicesUrl, walletName }, { keyAgent, stores });
+    this.#api$.next(wallet);
+  }
+
+  async deactivate(): Promise<void> {
+    this.#deactivateWallet();
+  }
+
+  shutdown(): void {
+    this.#deactivateWallet();
+
+    this.#hostSubscription.shutdown();
+    this.#api$.complete();
+    this.#walletStores.clear();
+  }
+
+  /** Resolves to the currently active wallet or waits until a wallet is activated. */
+  async walletActive(): Promise<ObservableWallet> {
+    return firstValueFrom(this.#api$.pipe(filter(isNotNil)));
+  }
+
+  async clearStore(walletName: string): Promise<void> {
+    if (!(await this.#isActive(walletName))) {
+      this.#walletStores.delete(walletName);
+    }
+  }
+
+  /** Gets store if wallet was activated previously or creates one when wallet is activated for the first time. */
+  #getStores(walletName: string): storage.WalletStores {
+    let stores = this.#walletStores.get(walletName);
+    if (!stores) {
+      stores = this.#storesFactory.create({ walletName });
+      this.#walletStores.set(walletName, stores);
+    }
+    return stores;
+  }
+
+  async #isActive(walletName: string): Promise<boolean> {
+    return (await this.#api$.value?.getName()) === walletName;
+  }
+
+  #deactivateWallet(): void {
+    const wallet = this.#api$?.getValue();
+    this.#api$.next(null);
```

Comment:
> * Would be better to not emit `null` before emitting new wallet if this is called in `activate`

The reason for emitting null before activating a new wallet is to avoid subscribers getting values from a wallet that is being `shutdown()`, before being switched to the new wallet. `null` is not actually seen by the clients. It's just a way to stop watching the wallet being deactivated.

Reply:
Fixed in [6f20bd1](https://github.com/input-output-hk/cardano-js-sdk/pull/502/commits/6f20bd189c069c7e3254dc989bd1cf3d055a0c4e)